### PR TITLE
Color Picker grabbag

### PR DESCRIPTION
**Pull request by volfied**
- Moved coalescing functionality into color picker
 - Add focus state highlight to color picker swatch so when it's tabbed to, it's highlighted, and toggles dialog with Enter/Esc
 - On dialog open, focus on the ColorInput, and tab becomes locked into the color picker dialog. Tab on Opacity input goes to Color input, and shift tab on color input goes to opacity input.
 - Fix a stroke issue #2382

### CODE REVIEW

**Comment by iwehrman** [comment]
Looks like the opacity input isn't hooked up?

**Comment by iwehrman** [comment]
The text input doesn't preserve the format:
purple

**Comment by iwehrman** [comment]
When you dismiss the picker with enter/escape, focus seems to return to the swatch but it does _not_ have the focused style.

**Comment by iwehrman** [comment]
Also, the swatch itself might need to be broken out into it's own component so that it can use the `Focusable` mixin. We need to make sure keyboard focus is returned to PS if you hit escape when the swatch is focused, for example.

**Comment by iwehrman** [comment]
Something weird is happening where if the swatch has focus and shows the focus ring, and then you cmd-tab away and back to PS, the focus ring disappears. This is probably related to the comment above about the focus ring not showing up when the dialog is dismissed.

File: src/js/jsx/shared/ColorPicker.jsx [hunk 1]
```diff
@@ -658,29 +813,91 @@ define(function (require, exports, module) {
...
+         * Focuses on the ColorType component
+         */
+        focusInput: function () {
+            this.refs.input.focus();
+        },
+
+        /**
+         * Focuses on the opacity input component
+         */
+        focusOpacityInput: function () {
```
Comment (iwehrman) [inline src/js/jsx/shared/ColorPicker.jsx]: Should be private?

**Comment by iwehrman** [comment]
Looks great besides a few issues with the swatch and the opacity input. Also note that I am very happy to merge this sooner rather than later and to deal with issues in follow up PRs. Just let me know.

**Comment by iwehrman** [comment]
In particular, I would merge this if you _only_ fixed the opacity input. But, if you'd rather try to fix up the swatch focus issues now, that's fine too.

**Comment by volfied** [comment]
@iwehrman Fixed the opacity box... As far as the focus goes, that is the problem I asked about this morning in the chat room. Using Focusable might be a viable option, but I'd like to keep that a separate issue as it's not a usability blocker.

**Comment by volfied** [comment]
Added two more changes based on @placegraphichere's comments.

**Comment by iwehrman** [comment]
OK, looks great! I'll file a few follow-up for some remaining issues that can wait until next week.

**Pull request merged**